File: src/main/java/me/chell/samsara/impl/gui/buttons/BindButton.java
package me.chell.samsara.impl.gui.buttons;

import me.chell.samsara.api.gui.GuiTheme;
import me.chell.samsara.api.value.Bind;
import me.chell.samsara.api.value.Value;
import me.chell.samsara.impl.gui.ValueButton;
import net.minecraft.client.util.InputUtil;
import net.minecraft.util.Formatting;
import org.lwjgl.glfw.GLFW;

public class BindButton extends ValueButton<Bind> {
    private boolean listening = false;

    public BindButton(Value<Bind> value, int x, int y) {
        super(x, y, GuiTheme.width, 13, value);
    }

    @Override
    public void draw(int mouseX, int mouseY) {
        drawThemedRectSecondary(x+2, y, width-2, height-1);
        drawThemedRectTertiary(x, y, 2, height);
        drawThemedRectTertiary(x+2, y+height-1, width-2, 1);

        drawThemedString(value.getDisplayName(), x + 4, getStringCenterY(y, height-1));

        String key = InputUtil.fromKeyCode(value.getValue().getKey(), 0).getLocalizedText().asString().toUpperCase();
        String text = (value.getValue().getHold() ? "HOLD " : "") + (key.isBlank() ? "UNKNOWN" : key);
        drawThemedStringRight(Formatting.GRAY + (listening ? "..." : text), x + width - 2, getStringCenterY(y, height-1));
    }

    @Override
    public boolean mouseClicked(int mouseX, int mouseY, int mouseButton) {
        if(mouseX >= x+2 && mouseX <= x + width && mouseY >= y && mouseY <= y + height-1) {
            if(listening) {
                value.getValue().setKey(mouseButton);
                listening = false;
            }
            else if(mouseButton == 0) {
                listening = true;
                return true;
            }
            else if(mouseButton == 1) {
                value.getValue().setHold(!value.getValue().getHold());
            }
        }
        return super.mouseClicked(mouseX, mouseY, mouseButton);
    }

    @Override
    public boolean keyTyped(char typedChar, int keyCode) {
        if(listening) {
            switch (keyCode) {
                case GLFW.GLFW_KEY_DELETE:
                    value.getValue().setKey(-1);
                    break;
                case GLFW.GLFW_KEY_ESCAPE:
                    break;
                default:
                    value.getValue().setKey(keyCode);
                    break;
            }
            listening = false;
            return true;
        }
        return super.keyTyped(typedChar, keyCode);
    }

    @Override
    public void onGuiClosed() {
        listening = false;
    }
}

File: src/main/kotlin/me/chell/samsara/impl/gui/click/buttons/value/BindButton.kt
package me.chell.samsara.impl.gui.click.buttons.value

import me.chell.samsara.api.gui.Rectangle
import me.chell.samsara.api.util.InputUtils
import me.chell.samsara.api.value.Bind
import me.chell.samsara.api.value.Value
import me.chell.samsara.impl.gui.click.Button
import me.chell.samsara.impl.gui.click.Window
import me.chell.samsara.impl.gui.click.buttons.ValueButton
import net.minecraft.client.util.math.MatrixStack
import org.lwjgl.glfw.GLFW

class BindButton(val v: Value<Bind>, override var x: Double, override var y: Double): ValueButton<Bind>(v, x, y) {

    private var listening = false

    override fun render(matrices: MatrixStack, mouseX: Double, mouseY: Double, tickDelta: Float) {
        val rect = Rectangle(x, y, Window.width, height).subtract(border)
        rect.x += Window.padding.left + Button.border.left
        rect.width -= Window.padding.left + Window.padding.right + Button.border.left + Button.border.right

        // draw background
        fill(matrices, rect, if(listening) primaryColor else secondaryColor)
        //draw border
        drawBorder(matrices, border, rect, borderColor)
        // draw text
        drawString(matrices, value.displayName, primaryText, rect)

        // draw value text
        val w = fontRenderer.getWidth(value.displayName)
        rect.x += w
        rect.width -= w

        var string = if(value.value.hold) "Hold " else ""
        string += InputUtils.getName(value.value.key)
        drawString(matrices, string, secondaryText, rect)
    }

    override fun keyPressed(keyCode: Int, scanCode: Int, modifiers: Int): Boolean {
        if(listening) {
            when(keyCode) {
                GLFW.GLFW_KEY_ESCAPE -> {
                    listening = false
                    return true
                }
                GLFW.GLFW_KEY_DELETE -> {
                    value.value.key = -1
                    listening = false
                    return true
                }
                else -> {
                    value.value.key = keyCode
                    listening = false
                    return true
                }
            }
        }
        return false
    }

    override fun mouseClicked(mouseX: Double, mouseY: Double, button: Int): Boolean {
        if(Rectangle(x, y, Window.width, height).isInBounds(mouseX, mouseY)) {
            when(button) {
                0 -> {
                    return if(listening) {
                        value.value.key = 0
                        listening = false
                        true
                    } else {
                        listening = true
                        true
                    }
                }
                1 -> {
                    return if (listening) {
                        value.value.key = 1
                        listening = false
                        true
                    } else {
                        value.value.hold = !value.value.hold
                        true
                    }
                }
                else -> {
                    if (listening) {
                        value.value.key = button
                        listening = false
                        return true
                    }
                }
            }
        }
        return false
    }
}